### PR TITLE
Rao-Blackwell adaptive MCMC

### PR DESCRIPTION
**Pull request by ben18785**
Addresses #917

### CODE REVIEW

**Comment by ben18785** [comment]
@MichaelClerx Any idea about this Travis error? Unexpected section title (referring to the 'references' bit of the docstrings).

**Comment by MichaelClerx** [comment]
For whatever reason it's not expecting a section title. Presumably it's still inside some other structure, like a list of bullet points or a code block or something? Should go away if you tidy up the bit above, add in the proper latex tags etc.

**Comment by ben18785** [comment]
@MichaelClerx Thanks!

File: pints/_mcmc/_rao_blackwell_ac.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#
+# Rao-Blackwell Adaptive MCMC method
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2018, University of Oxford.
```
Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: 2019

File: pints/_mcmc/_rao_blackwell_ac.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import pints
+import numpy as np
+
+
+class RaoBlackwellACMC(pints.GlobalAdaptiveCovarianceMC):
+    """
+    Rao-Blackwell adaptive MCMC, as described by Algorithm 3 in [1]_. The
+    algorithnm consists of the following steps::
```
Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: `::`, on its own, indicates the start of a code block. But "initialise mu0 and sigma0" is not code

File: pints/_mcmc/_rao_blackwell_ac.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        Initialise mu0 and sigma0
+
+    In each iteration after initial phase (t)::
+
+        theta* ~ N(theta_t, lambda * Sigma0)
```
Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: Is this meant to be pseudo-code or latex??

Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: There's a few `Sigma0`s instead of `sigma0`

File: pints/_mcmc/_rao_blackwell_ac.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+            sigma_t+1 = sigma_t + gamma_t+1 *
+                        (bar((theta_t+1 - mu_t)(theta_t+1 - mu_t)') - sigma_t)
+
+        where::
+
+            bar(X_t+1) = alpha(X_t, Y_t+1) * Y_t+1 +
+                            (1 - alpha(X_t, Y_t+1)) * X_t
+        """
+        _acceptance_prob = (
```
Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: Why the leading underscore?

File: pints/_mcmc/_rao_blackwell_ac.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    ----------
+    .. [1] A tutorial on adaptive MCMC
+           Christophe Andrieu and Johannes Thoms, Statistical Computing, 2008,
+           18: 343-373.
+           https://doi.org/10.1007/s11222-008-9110-y
+    """
+    def __init__(self, x0, sigma0=None):
+        super(RaoBlackwellACMC, self).__init__(x0, sigma0)
+
+        self._lambda = (2.38**2) / self._n_parameters
```
Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: Is this a heuristic?

File: pints/_mcmc/_rao_blackwell_ac.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        """ See :meth:`TunableMethod.n_hyper_parameters()`. """
+        return 1
+
+    def name(self):
+        """ See :meth:`pints.MCMCSampler.name()`. """
+        return 'Rao-Blackwell adaptive covariance MCMC'
+
+    def tell(self, fx):
+        """ See :meth:`pints.AdaptiveCovarianceMCMC.tell()`. """
+        self._Y = np.copy(self._proposed)
```
Comment (MichaelClerx) [inline pints/_mcmc/_rao_blackwell_ac.py]: These should also be declared in the constructor

File: examples/sampling-adaptive-covariance-rao-blackwell.ipynb [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Inference: Adaptive Covariance Rao-Blackwell MCMC\n",
+    "\n",
+    "This example shows you how to perform Bayesian inference on a time series, using a variant of [Adaptive Covariance MCMC](http://pints.readthedocs.io/en/latest/mcmc_samplers/adaptive_covariance_mcmc_remi.html) detailed in Algorithm 3 of [1].\n",
+    "\n",
+    "[1] A tutorial on adaptive MCMC\n",
```
Comment (MichaelClerx) [inline examples/sampling-adaptive-covariance-rao-blackwell.ipynb]: Make this a link to https://doi.org/10.1007/s11222-008-9110-y ?

File: examples/sampling-adaptive-covariance-rao-blackwell.ipynb [hunk 8]
```diff
@@ -0,0 +1,150 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Inference: Adaptive Covariance Rao-Blackwell MCMC\n",
+    "\n",
+    "This example shows you how to perform Bayesian inference on a time series, using a variant of [Adaptive Covariance MCMC](http://pints.readthedocs.io/en/latest/mcmc_samplers/adaptive_covariance_mcmc_remi.html) detailed in Algorithm 3 of [1].\n",
```
Comment (MichaelClerx) [inline examples/sampling-adaptive-covariance-rao-blackwell.ipynb]: Link should point here: https://pints.readthedocs.io/en/latest/mcmc_samplers/adaptive_covariance_mc.html

File: examples/sampling-adaptive-covariance-rao-blackwell.ipynb [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
...WKl6LvxHlUqNuMWRPH0K9jY3ZsXJdg0sydrwA9Br7JtOVb6dJ/KH/t3cnTHZrxZLsm7Ai9\n84pIROj9wmDe/2Qq2zatp33r5ly7qt8q040Lf7lufizTwupI4nTIFGadoyJd/VaBI9rqcFK15M4u\n+xbX8v6jcK2I3NYYMzclA0vLBg4eRqbMQXz+8eg73ntj+Pv4+fkxa+KYeI/v/NwrVG3anjVff8q+\n9av+j73zDovi6uLwe3eX3gVUEFTsPfbee+/dxN6Nny12jcaWGGONRmOJvSXG3nuJDXtv2BULqCBK\nXfZ+f+yi4C6CWBZh3ufZh92ZO3fOLDBn7r3n/E687YrUaoatqweXNsw1uoHbZcjJq3sX3ntjl1Ly\n9Mgazk9qxs0lg7m1fBhXfm/PhRldCLltnDTpkKUwWVuPJ+LZA/ZOG4ROF41LppyUGzCTiJDnzB3U\nifDX+hu9tb0DzYdPpdWPvxP+OoQfuzanR9OqbF296L1qAfaOTjTv2pf520/QedBYHtz2o22janTv\n2Ar/B/eM2tdt3JKJsxZx/vQJWjVvrEydpRSuGUQ7ctYyrx3vYXl0VTzEc/CL/39UIenRZUullFel\nlLOklDOllFeEEEs/tderiUgAACAASURBVHFfK27u7nTq2oOdW9YZJR6mTe9B6w7dOLR9PQ/vmF7o\nFkIwfPwUMuTIz5qJAwkOMB3FotZYUK1tL57fumg0mslavDxRwU8JD4g/Wi3QdwMPd8xG45YVxyqD\ncKo9FrtibdFFvsJv0QACjhsnlzpkKYRX7e8J8TvB9e36X7lrlnyU6jmR4Ic3mTu0O9Gxnuzylq9J\nn4U7qPe/0WijIpkzfggdqhakU90y/DF2EPs2rzEZJGBlY6tfs9lyjNa9BnHy0B4aVi7GqkXGDrV6\nnYb89NsfHP9vP106tleCAVIC17ZBunzgkmCJKrOxR1eIAOkEpxeb25RkTVKny+LonxjWZ4p8vDkp\nh87dewHw70rjP8BvO/VCrdaw7e/4/zgtLK34ceo8orVRbJsd/0xk4RqNsXX14PLGeXG2e3xTDoDg\nq0dMHhcdGcbDPYuw8MiHY9XBWHoXxiJtdmzy1salwSQsvArzYPscIp49NDrWtUgdnHKV5uL6uYS+\n0Ac4eBQoQ9H2I3h62ZeFYwbFudFbWFpRon4b
```
Comment (MichaelClerx) [inline examples/sampling-adaptive-covariance-rao-blackwell.ipynb]: One of these notebooks should maybe mention the sudden convergence after the initial phase. But explain why we still don't just start off adapting

File: pints/_mcmc/_rao_blackwell_ac.py (hunk 1 above)
Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: Done

File: pints/_mcmc/_rao_blackwell_ac.py (hunk 2 above)
Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: Changed

File: pints/_mcmc/_rao_blackwell_ac.py (hunk 3 above)
Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: Pseudocode

Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: Changed

File: pints/_mcmc/_rao_blackwell_ac.py (hunk 4 above)
Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: changed

File: pints/_mcmc/_rao_blackwell_ac.py (hunk 5 above)
Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: Yes, but based on normal approximation.

File: pints/_mcmc/_rao_blackwell_ac.py (hunk 6 above)
Comment (ben18785) [reply pints/_mcmc/_rao_blackwell_ac.py]: Done

File: examples/sampling-adaptive-covariance-rao-blackwell.ipynb (hunk 7 above)
Comment (ben18785) [reply examples/sampling-adaptive-covariance-rao-blackwell.ipynb]: Done

File: examples/sampling-adaptive-covariance-rao-blackwell.ipynb (hunk 8 above)
Comment (ben18785) [reply examples/sampling-adaptive-covariance-rao-blackwell.ipynb]: Done

File: examples/sampling-adaptive-covariance-rao-blackwell.ipynb (hunk 9 above)
Comment (ben18785) [reply examples/sampling-adaptive-covariance-rao-blackwell.ipynb]: Added some notes about this; good point.

**Pull request merged**